File: src/session/codex.ts
import { createHash } from "node:crypto";
import {
  closeSync,
  existsSync,
  openSync,
  readFileSync,
  readSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { join, relative } from "node:path";
import { shq } from "../util/shell.ts";
import type { Transport } from "../transport/types.ts";
import {
  cleanupGuardedHomeFile,
  collectGuardedHomeFile,
  installGuardedHomeFile,
} from "./guarded-store.ts";
import type {
  InstalledSession,
  InstallOptions,
  LocalSession,
  SessionAdapter,
  StagedReturn,
} from "./types.ts";

/**
 * Codex stores sessions as ~/.codex/sessions/YYYY/MM/DD/rollout-<ts>-<id>.jsonl.
 * The first line is a `session_meta` record whose payload carries the session
 * id and cwd. Resume is `codex resume <id>` from any directory, so beam keeps
 * the same home-relative store path on the target.
 */

const CANDIDATE_SCAN_COUNT = 400;

/**
 * Read only the leading bytes of a transcript and return its first line —
 * the `session_meta` record. locate inspects up to CANDIDATE_SCAN_COUNT
 * candidates and transcripts grow to many megabytes, so no candidate is
 * ever read whole.
 * A first line longer than the cap comes back truncated, parses as malformed
 * JSON, and the candidate is skipped — the same outcome as any other corrupt
 * header. Read errors propagate, exactly as a whole-file read's would.
 */
export const HEADER_SCAN_BYTES = 64 * 1024;

function readHeaderLine(file: string): string {
  let fd: number | undefined;
  try {
    fd = openSync(file, "r");
    const buf = Buffer.alloc(HEADER_SCAN_BYTES);
    const n = readSync(fd, buf, 0, buf.length, 0);
    return buf.toString("utf8", 0, n).split("\n", 1)[0] ?? "";
  } finally {
    if (fd !== undefined) closeSync(fd);
  }
}

interface SessionMeta {
  type?: string;
  payload?: { session_id?: string; id?: string; cwd?: string };
}
function assertCodexTranscript(text: string, sessionId: string, expectedCwd?: string): void {
  const firstLine = text.split("\n", 1)[0] ?? "";
  let meta: SessionMeta;
  try {
    meta = JSON.parse(firstLine) as SessionMeta;
  } catch {
    throw new Error("remote Codex transcript contains invalid session metadata");
  }
  const identity = meta.payload?.session_id ?? meta.payload?.id;
  if (meta.type !== "session_meta" || identity !== sessionId) {
    throw new Error(`remote Codex transcript does not belong to session ${sessionId}`);
  }
  if (expectedCwd !== undefined && meta.payload?.cwd !== expectedCwd) {
    throw new Error(
      `remote Codex transcript records cwd ${meta.payload?.cwd ?? "(none)"}, ` +
      `not the shipped workspace ${expectedCwd}`,
    );
  }
}

export class CodexAdapter implements SessionAdapter {
  readonly tool = "codex" as const;
  readonly binary = "codex";
  readonly loginArgv = ["codex", "login"];
  readonly remoteAuthProbe = 'test -s "$HOME/.codex/auth.json"';

  async locate(cwd: string, home: string, sessionRef?: string): Promise<LocalSession | undefined> {
    const root = join(home, ".codex", "sessions");
    if (!existsSync(root)) return undefined;
    const files: { file: string; mtime: number }[] = [];
    for (const rel of new Bun.Glob("*/*/*/rollout-*.jsonl").scanSync({ cwd: root })) {
      const file = join(root, rel);
      files.push({ file, mtime: statSync(file).mtimeMs });
    }
    files.sort((a, b) => b.mtime - a.mtime);

    for (const { file, mtime } of files.slice(0, CANDIDATE_SCAN_COUNT)) {
      const firstLine = readHeaderLine(file);
      let meta: SessionMeta;
      try {
        meta = JSON.parse(firstLine) as SessionMeta;
      } catch {
        continue;
      }
      if (meta.type !== "session_meta" || meta.payload?.cwd !== cwd) continue;
      const id = meta.payload.session_id ?? meta.payload.id;
      if (!id) continue;
      if (sessionRef && !id.startsWith(sessionRef)) continue;
      return { tool: "codex", id, file, mtime };
    }
    return undefined;
  }

  /** Home-relative store segments, reused verbatim on the target. */
  private storePath(source: string): string[] {
    const home = source.split("/.codex/")[0]!;
    return relative(home, source).split("/");
  }

  async install(
    t: Transport,
    session: LocalSession,
    _remoteCwd: string,
    { kickoff }: InstallOptions = {},
  ): Promise<InstalledSession> {
    // `file` is the staged CONTENT source; the remote layout mirrors the
    // LIVE store path (a ship-stage path has no `/.codex/` component).
    const path = this.storePath(session.storeFile ?? session.file);
    assertCodexTranscript(readFileSync(session.file, "utf8"), session.id);
    const remoteStore = await installGuardedHomeFile(t, session.file, path);
    return {
      resumeArgv: ["codex", "resume", session.id, ...(kickoff ? [kickoff] : [])],
      notes: [
        `session -> ${remoteStore}`,
        "note: codex records the original cwd in session_meta; it resumes in the current directory",
      ],
    };
  }

  async stageReturn(
    t: Transport,
    session: LocalSession,
    localCwd: string,
    _remoteCwd: string,
    stageDir: string,
  ): Promise<StagedReturn> {
    const path = this.storePath(session.storeFile ?? session.file);
    const returned = await collectGuardedHomeFile(t, path);
    assertCodexTranscript(returned, session.id, localCwd);
    // Codex resumes by id from its own store only — the hint is the exact
    // manual import; beam itself never writes into the live ~/.codex store.
    writeFileSync(join(stageDir, "session.jsonl"), returned);
    const hint =
      `manual import (codex cannot resume an isolated path): ` +
      `cp ${shq(join(stageDir, "session.jsonl"))} ` +
      `${shq(session.file)} && codex resume ${shq(session.id)} ` +
      `# replaces your local copy of this session; it was left untouched`;
    return {
      hint,
      remoteSessionSha256: createHash("sha256").update(returned).digest("hex"),
    };
  }

  async cleanupRemote(t: Transport, session: LocalSession, _remoteCwd: string): Promise<void> {
    await cleanupGuardedHomeFile(t, this.storePath(session.storeFile ?? session.file));
  }
}

File: rust/src/session/codex.rs
//! Codex CLI session adapter.

use std::fs::{self, File};
use std::io::Read;
use std::path::{Component, Path, PathBuf};

use serde::Deserialize;

use crate::session::claude::path_text;
use crate::session::guarded_store::{
    cleanup_guarded_home_file, collect_guarded_home_file, install_guarded_home_file,
};
use crate::session::{
    InstallOptions, InstalledSession, LocalSession, SessionAdapter, SessionError, SessionFuture,
    StagedReturn, ToolName,
};
use crate::transport::Transport;
use crate::util::digest::file_sha256;
use crate::util::shell::shq;

const CANDIDATE_SCAN_COUNT: usize = 400;
pub const HEADER_SCAN_BYTES: usize = 64 * 1024;
const MAX_CODEX_STORE_ENTRIES: usize = 65_536;
const MAX_CODEX_STORE_SEGMENTS: usize = 16;

pub struct CodexAdapter;

#[derive(Deserialize)]
struct SessionMeta {
    #[serde(rename = "type")]
    kind: Option<String>,
    payload: Option<SessionPayload>,
}

#[derive(Deserialize)]
struct SessionPayload {
    session_id: Option<String>,
    id: Option<String>,
    cwd: Option<String>,
}

fn read_header_line(file: &Path) -> Result<String, SessionError> {
    let source = File::open(file).map_err(|error| {
        SessionError::caused_by(
            format!("cannot open Codex transcript {}", file.display()),
            error,
        )
    })?;
    let mut bytes = Vec::with_capacity(HEADER_SCAN_BYTES);
    source
        .take(HEADER_SCAN_BYTES as u64)
        .read_to_end(&mut bytes)
        .map_err(|error| {
            SessionError::caused_by(
                format!("cannot read Codex transcript header {}", file.display()),
                error,
            )
        })?;
    let text = String::from_utf8_lossy(&bytes);
    Ok(text.split('\n').next().unwrap_or("").to_owned())
}

fn parse_meta(text: &str) -> Result<SessionMeta, serde_json::Error> {
    let first_line = text.split('\n').next().unwrap_or("");
    serde_json::from_str(first_line)
}

fn assert_codex_transcript(
    text: &str,
    session_id: &str,
    expected_cwd: Option<&str>,
) -> Result<(), SessionError> {
    let meta = parse_meta(text).map_err(|source| {
        SessionError::caused_by(
            "remote Codex transcript contains invalid session metadata".to_owned(),
            source,
        )
    })?;
    let payload = meta.payload.as_ref();
    let identity = payload.and_then(|value| value.session_id.as_ref().or(value.id.as_ref()));
    if meta.kind.as_deref() != Some("session_meta")
        || identity.map(String::as_str) != Some(session_id)
    {
        return Err(SessionError::message(format!(
            "remote Codex transcript does not belong to session {session_id}"
        )));
    }
    if let Some(expected_cwd) = expected_cwd {
        let recorded = payload.and_then(|value| value.cwd.as_deref());
        if recorded != Some(expected_cwd) {
            return Err(SessionError::message(format!(
                "remote Codex transcript records cwd {}, not the shipped workspace {expected_cwd}",
                recorded.unwrap_or("(none)")
            )));
        }
    }
    Ok(())
}

fn codex_store_path(source: &Path) -> Result<Vec<String>, SessionError> {
    let mut found = false;
    let mut segments = Vec::new();
    for component in source.components() {
        match component {
            Component::Normal(value) => {
                let text = value.to_str().ok_or_else(|| {
                    SessionError::message(format!(
                        "Codex transcript path {} is not valid UTF-8",
                        source.display()
                    ))
                })?;
                if !found {
                    if text != ".codex" {
                        continue;
                    }
                    found = true;
                }
                segments.push(text.to_owned());
                if segments.len() > MAX_CODEX_STORE_SEGMENTS {
                    return Err(SessionError::message(format!(
                        "Codex transcript store path exceeds {MAX_CODEX_STORE_SEGMENTS} components"
                    )));
                }
            }
            Component::CurDir | Component::ParentDir => {
                if found {
                    return Err(SessionError::message(format!(
                        "Codex transcript path {} has an unsafe store component",
                        source.display()
                    )));
                }
            }
            Component::RootDir | Component::Prefix(_) => {
                if found {
                    return Err(SessionError::message(format!(
                        "Codex transcript path {} has an unsafe store root",
                        source.display()
                    )));
                }
            }
        }
    }
    if !found || segments.len() < 2 {
        return Err(SessionError::message(format!(
            "Codex transcript {} is not inside a .codex store",
            source.display()
        )));
    }
    Ok(segments)
}

fn collect_rollout_files(
    root: &Path,
) -> Result<Vec<(PathBuf, std::time::SystemTime)>, SessionError> {
    let mut files = Vec::new();
    let mut pending = vec![(root.to_path_buf(), 3_usize)];
    let mut walked = 0_usize;
    while let Some((directory, levels_remaining)) = pending.pop() {
        let metadata = match fs::metadata(&directory) {
            Ok(metadata) => metadata,
            Err(source) if source.kind() == std::io::ErrorKind::NotFound => continue,
            Err(source) => {
                return Err(SessionError::caused_by(
                    format!("cannot inspect Codex session path {}", directory.display()),
                    source,
                ));
            }
        };
        if !metadata.is_dir() {
            continue;
        }
        let entries = fs::read_dir(&directory).map_err(|source| {
            SessionError::caused_by(
                format!("cannot read Codex session path {}", directory.display()),
                source,
            )
        })?;
        for entry in entries {
            walked += 1;
            if walked > MAX_CODEX_STORE_ENTRIES {
                return Err(SessionError::message(format!(
                    "Codex session store exceeds {MAX_CODEX_STORE_ENTRIES} entries — narrow it \
                     before retrying"
                )));
            }
            let entry = entry.map_err(|source| {
                SessionError::caused_by(
                    format!("cannot scan Codex session path {}", directory.display()),
                    source,
                )
            })?;
            if levels_remaining > 0 {
                pending.push((entry.path(), levels_remaining - 1));
                continue;
            }
            let name = entry.file_name();
            let Some(name) = name.to_str() else {
                continue;
            };
            if !name.starts_with("rollout-") || !name.ends_with(".jsonl") {
                continue;
            }
            let modified = entry
                .metadata()
                .and_then(|value| value.modified())
                .map_err(|source| {
                    SessionError::caused_by(
                        format!("cannot inspect Codex transcript {}", entry.path().display()),
                        source,
                    )
                })?;
            files.push((entry.path(), modified));
        }
    }
    files.sort_by_key(|file| std::cmp::Reverse(file.1));
    Ok(files)
}

impl SessionAdapter for CodexAdapter {
    fn tool(&self) -> ToolName {
        ToolName::Codex
    }

    fn binary(&self) -> &'static str {
        "codex"
    }

    fn login_argv(&self) -> &'static [&'static str] {
        &["codex", "login"]
    }

    fn remote_auth_probe(&self) -> Option<&'static str> {
        Some(r#"test -s "$HOME/.codex/auth.json""#)
    }

    fn locate<'a>(
        &'a self,
        cwd: &'a Path,
        home: &'a Path,
        session_ref: Option<&'a str>,
    ) -> SessionFuture<'a, Option<LocalSession>> {
        Box::pin(async move {
            let root = home.join(".codex").join("sessions");
            if !root.exists() {
                return Ok(None);
            }
            let cwd = path_text(cwd, "Codex workspace")?;
            for (file, modified) in collect_rollout_files(&root)?
                .into_iter()
                .take(CANDIDATE_SCAN_COUNT)
            {
                let header = read_header_line(&file)?;
                let Ok(meta) = parse_meta(&header) else {
                    continue;
                };
                let Some(payload) = meta.payload else {
                    continue;
                };
                if meta.kind.as_deref() != Some("session_meta") {
                    continue;
                }
                if payload.cwd.as_deref() != Some(cwd) {
                    continue;
                }
                let Some(id) = payload.session_id.or(payload.id) else {
                    continue;
                };
                if session_ref.is_some_and(|reference| !id.starts_with(reference)) {
                    continue;
                }
                return Ok(Some(LocalSession {
                    tool: ToolName::Codex,
                    id,
                    file,
                    store_file: None,
                    artifacts_dir: None,
                    modified,
                }));
            }
            Ok(None)
        })
    }

    fn install<'a>(
        &'a self,
        transport: &'a dyn Transport,
        session: &'a LocalSession,
        _remote_cwd: &'a str,
        options: InstallOptions<'a>,
    ) -> SessionFuture<'a, InstalledSession> {
        Box::pin(async move {
            let header = read_header_line(&session.file)?;
            assert_codex_transcript(&header, &session.id, None)?;
            let source = session.store_file.as_deref().unwrap_or(&session.file);
            let path = codex_store_path(source)?;
            let references = path.iter().map(String::as_str).collect::<Vec<_>>();
            let remote_store =
                install_guarded_home_file(transport, &session.file, &references).await?;
            let mut resume_argv = vec!["codex".to_owned(), "resume".to_owned(), session.id.clone()];
            if let Some(kickoff) = options.kickoff {
                resume_argv.push(kickoff.to_owned());
            }
            Ok(InstalledSession {
                resume_argv,
                notes: vec![
                    format!("session -> {remote_store}"),
                    "note: codex records the original cwd in session_meta; it resumes in the current \
                     directory"
                        .to_owned(),
                ],
            })
        })
    }

    fn stage_return<'a>(
        &'a self,
        transport: &'a dyn Transport,
        session: &'a LocalSession,
        local_cwd: &'a Path,
        _remote_cwd: &'a str,
        stage_dir: &'a Path,
    ) -> SessionFuture<'a, StagedReturn> {
        Box::pin(async move {
            let source = session.store_file.as_deref().unwrap_or(&session.file);
            let path = codex_store_path(source)?;
            let references = path.iter().map(String::as_str).collect::<Vec<_>>();
            let returned = collect_guarded_home_file(transport, &references).await?;
            let header = read_header_line(returned.path())?;
            let local_cwd = path_text(local_cwd, "Codex local workspace")?;
            assert_codex_transcript(&header, &session.id, Some(local_cwd))?;
            let staged = stage_dir.join("session.jsonl");
            fs::copy(returned.path(), &staged).map_err(|source| {
                SessionError::caused_by(
                    format!("cannot write staged Codex return {}", staged.display()),
                    source,
                )
            })?;
            let hint = format!(
                "manual import (codex cannot resume an isolated path): cp {} {} && codex resume {} \
                 # replaces your local copy of this session; it was left untouched",
                shq(path_text(&staged, "Codex staged return")?),
                shq(path_text(&session.file, "Codex local transcript")?),
                shq(&session.id)
            );
            Ok(StagedReturn {
                hint,
                remote_session_sha256: file_sha256(returned.path()).map_err(|source| {
                    SessionError::caused_by(
                        "cannot hash returned Codex transcript".to_owned(),
                        source,
                    )
                })?,
            })
        })
    }

    fn cleanup_remote<'a>(
        &'a self,
        transport: &'a dyn Transport,
        session: &'a LocalSession,
        _remote_cwd: &'a str,
    ) -> SessionFuture<'a, ()> {
        Box::pin(async move {
            let source = session.store_file.as_deref().unwrap_or(&session.file);
            let path = codex_store_path(source)?;
            let references = path.iter().map(String::as_str).collect::<Vec<_>>();
            cleanup_guarded_home_file(transport, &references, false).await
        })
    }
}
